File: truevfs-kernel/src/main/java/de/schlichtherle/truevfs/kernel/se/ReadOnlyArchiveFileSystem.java
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truevfs.kernel.se;

import net.truevfs.kernel.*;
import net.truevfs.kernel.cio.Container;
import net.truevfs.kernel.cio.Entry;
import net.truevfs.kernel.cio.Entry.Access;
import static net.truevfs.kernel.cio.Entry.Access.READ;
import net.truevfs.kernel.util.BitField;
import java.io.IOException;
import java.util.Map;
import javax.annotation.CheckForNull;
import javax.annotation.WillNotClose;
import javax.annotation.concurrent.NotThreadSafe;

/**
 * A read-only virtual file system for archive entries.
 * <p>
 * All modifying methods throw a {@link FsReadOnlyFileSystemException}.
 *
 * @param  <E> The type of the archive entries.
 * @author Christian Schlichtherle
 */
@NotThreadSafe
final class ReadOnlyArchiveFileSystem<E extends FsArchiveEntry>
extends ArchiveFileSystem<E> {

    private static final BitField<Access> READ_ONLY = BitField.of(READ);

    ReadOnlyArchiveFileSystem(
            final FsArchiveDriver<E> driver,
            final @WillNotClose Container<E> archive,
            final @CheckForNull Entry rootTemplate) {
        super(driver, archive, rootTemplate);
    }

    @Override
    void checkAccess(
            final BitField<FsAccessOption> options,
            final FsEntryName name,
            final BitField<Access> types)
    throws IOException {
        if (!types.isEmpty() && !READ_ONLY.equals(types))
            throw new FsReadOnlyFileSystemException();
        super.checkAccess(options, name, types);
    }

    @Override
    void setReadOnly(FsEntryName name) { }

    @Override
    boolean setTime(
            BitField<FsAccessOption> options,
            FsEntryName name,
            Map<Access, Long> times)
    throws IOException {
        throw new FsReadOnlyFileSystemException();
    }

    @Override
    boolean setTime(
            BitField<FsAccessOption> options,
            FsEntryName name,
            BitField<Access> types,
            long value)
    throws IOException {
        throw new FsReadOnlyFileSystemException();
    }

    @Override
    Mknod mknod(
            BitField<FsAccessOption> options,
            FsEntryName name,
            Entry.Type type,
            Entry template)
    throws IOException {
        throw new FsReadOnlyFileSystemException();
    }

    @Override
    void unlink(BitField<FsAccessOption> options, FsEntryName name)
    throws IOException {
        throw new FsReadOnlyFileSystemException();
    }
}

File: truevfs-kernel/src/main/scala/de/schlichtherle/truevfs/kernel/se/ReadOnlyArchiveFileSystem.scala
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truevfs.kernel.se

import net.truevfs.kernel._
import net.truevfs.kernel.cio._
import net.truevfs.kernel.cio.Entry._;
import net.truevfs.kernel.cio.Entry.Access._;
import net.truevfs.kernel.util._

/**
 * A read-only virtual file system for archive entries.
 * <p>
 * All modifying methods throw a {@link FsReadOnlyFileSystemException}.
 *
 * @param  <E> The type of the archive entries.
 * @author Christian Schlichtherle
 */
private final class ReadOnlyArchiveFileSystem[E <: FsArchiveEntry](
  driver: FsArchiveDriver[E],
  archive: Container[E],
  rootTemplate: Option[Entry])
extends ArchiveFileSystem(driver, archive, rootTemplate) {
  import ReadOnlyArchiveFileSystem._

  override def checkAccess(options: AccessOptions, name: FsEntryName, types: BitField[Access]) {
    if (!types.isEmpty && READ_ONLY != types)
        throw new FsReadOnlyFileSystemException();
    super.checkAccess(options, name, types)
  }

  override def setReadOnly(name: FsEntryName) { }

  override def setTime(options: AccessOptions, name: FsEntryName, times: Map[Access, Long]) =
    throw new FsReadOnlyFileSystemException

  override def setTime(options: AccessOptions, name: FsEntryName, types: BitField[Access], value: Long) =
    throw new FsReadOnlyFileSystemException

  override def mknod(options: AccessOptions, name: FsEntryName, tµpe: Type, template: Option[Entry]) =
    throw new FsReadOnlyFileSystemException

  override def unlink(options: AccessOptions, name: FsEntryName) =
    throw new FsReadOnlyFileSystemException
}

private object ReadOnlyArchiveFileSystem {
  private val READ_ONLY = BitField.of(READ)
}
